contracts/utils/cryptography/P256.sol

```diff
@@ -0,0 +1,304 @@
+    /// @dev B parameter of the weierstrass equation
+    uint256 internal constant B = 0x5AC635D8AA3A93E7B3EBBD55769886BC651D06B0CC53B0F63BCE3C3E27D2604B;
+
+    /// @dev (P + 1) / 4. Useful to compute sqrt
+    uint256 private constant P1DIV4 = 0x3fffffffc0000000400000000000000000000000400000000000000000000000;
+
+    /// @dev N/2 for excluding higher order `s` values
+    uint256 private constant HALF_N = 0x7fffffff800000007fffffffffffffffde737d56d38bcf4279dce5617e3192a8;
+
+    /**
+     * @dev Verifies a secp256r1 signature using the RIP-7212 precompile and falls back to the Solidity implementation
+     * if the precompile is not available. This version should work on all chains, but requires the deployment of more
+     * bytecode.
+     *
+     * @param h - hashed message
+     * @param r - signature half R
+     * @param s - signature half S
+     * @param qx - public key coordinate X
+     * @param qy - public key coordinate Y
+     *
+     * IMPORTANT: This function disallows signatures where the `s` value is above `N/2` to prevent malleability.
+     * To flip the `s` value, compute `s = N - s`.
+     */
+    function verify(bytes32 h, bytes32 r, bytes32 s, bytes32 qx, bytes32 qy) internal view returns (bool) {
+        (bool valid, bool supported) = _tryVerifyNative(h, r, s, qx, qy);
+        return supported ? valid : verifySolidity(h, r, s, qx, qy);
+    }
+
+    /**
+     * @dev Same as {verify}, but it will revert if the required precompile is not available.
+     */
+    function verifyNative(bytes32 h, bytes32 r, bytes32 s, bytes32 qx, bytes32 qy) internal view returns (bool) {
+        (bool valid, bool supported) = _tryVerifyNative(h, r, s, qx, qy);
+        if (supported) {
+            return valid;
+        } else {
+            revert Errors.MissingPrecompile(address(0x100));
+        }
+    }
+
+    /**
+     * @dev Same as {verify}, but it will return false if the required precompile is not available.
+     */
+    function _tryVerifyNative(
+        bytes32 h,
+        bytes32 r,
+        bytes32 s,
+        bytes32 qx,
+        bytes32 qy
+    ) private view returns (bool valid, bool supported) {
+        if (r == 0 || uint256(r) >= N || s == 0 || uint256(s) > HALF_N || !isOnCurve(qx, qy)) {
+            return (false, true); // signature is invalid, and its not because the precompile is missing
+        }
+
+        (bool success, bytes memory returndata) = address(0x100).staticcall(abi.encode(h, r, s, qx, qy));
+        return (success && returndata.length == 0x20) ? (abi.decode(returndata, (bool)), true) : (false, false);
+    }
+
+    /**
+     * @dev Same as {verify}, but only the Solidity implementation is used.
+     */
+    function verifySolidity(bytes32 h, bytes32 r, bytes32 s, bytes32 qx, bytes32 qy) internal view returns (bool) {
+        if (r == 0 || uint256(r) >= N || s == 0 || uint256(s) > HALF_N || !isOnCurve(qx, qy)) {
+            return false;
+        }
```

Comment:
for a public key to be valid, it should:
1. not be point of infinity
2. have coordinates $(x, y)$ satisfy $0 <= x, y <= p$, and
3. satisfy the equation $y^2=x3+ax+b$ modulo $p$

on property one: the public key is used in `_preComputeJacobianPoints` by `points[0x01] = JPoint(px, py, 1); // 1,0 (p)`, which would never be the point of infinity. `_affineFromJacobian` uses the convention that $(0,0)$ represents the point of infinity, so in this case, the conversion in `_preComputeJacobianPoints` (L278) would be incorrect, but given that the point of infinity is not valid anyways, it's not an issue as we instead reject the public key $(0,0)$ via `isOnCurve`.

on property two: x and y coordinates of the public key get reduced to modulo p in calculations, so missing a check for property two means that the verify function will check the signature for the public key with coordinates $(x \bmod p, y \bmod p)$. This would allow for public keys $(x,y)$, where $x<2^{256} - p$ or $y<2^{256} - p$, which there would be another pair $(x^\prime,y^\prime)$ — such as $(x+y,y)$ — that can be used as a public key and for which signatures made for $(x,y)$ would also verify.

on property three: we are checking that it lies on the curve via the Weierstrass equation (`isOnCurve`).

i think properties 1 and 3 are ok, but missing 2.

Reply:
So should we add `uint256(qx) >= P || uint256(qy) >= P` ? Also, is it P or N?

Reply:
yeah that should be good. it's P because we are using that for the modulus operations on places like `isOnCurve`  
```suggestion
    function verifySolidity(bytes32 h, bytes32 r, bytes32 s, bytes32 qx, bytes32 qy) internal view returns (bool) {
        if (r == 0 || uint256(r) >= N || s == 0 || uint256(s) > HALF_N || uint256(qx) >= P || uint256(qy) >= P || !isOnCurve(qx, qy)) {
            return false;
        }
```

tests pass with this change, but i think it's useful to add a unit test for this property specifically, i can add it

Reply:
i think it also makes sense to have this in `_tryVerifyNative`

Reply:
I added the check directly in `isOnCurve`

Reply:
Note: the public key is unlikelly to be arbitrary user input. It will be recorded/whitelisted onchain, by some admin. So this attack vector is not really critical. But still its good to add the (cheap) check in `isOnCurve`

Reply:
> This would allow for public keys $(x,y)$, where $x<2^{256} - p$ or $y<2^{256} - p$, which there would be another pair $(x^\prime,y^\prime)$ — such as $(x+p,y)$ — that can be used as a public key and for which signatures made for $(x,y)$ would also verify.'

Although this is true, it's unclear to me whether you can create a signature for public keys $(x,y)$ if it's not on the curve. I still think checking is valuable and pretty much agree with @Amxx's conclusion that mostly public keys won't be arbitrary and the check is cheap.

Reply:
`if it's not on the curve`: the problem is that you can find $(x^\prime,y^\prime)$ (like $(x+p,y)$ ) that are on the curve based on the original $(x,y)$ which are on the curve too -- it's like a signature malleability attack vector.

Reply:
Got it, so it's the other way around. Still it's relevant that public keys are unlikely to be arbitrary, but let's keep the check regardless.